### ISSUE TITLE
Considering Possibility for Future Collaboration? (edited issue title) 

### ISSUE THREAD

**Issue by Librechain** [antoniusf/encrypted-stream#1]
**Edit**: Amended this issue entirely after further reflection (as well as feedback from maintainer) has led me to reconsider my original (harsh) opinion. 

~~You have no clue what people were using the code for. It could have been as a boiler plate or for numerous other reasons.~~ 

~~But without asking, you decided that you knew better than any and all individuals that may or may not have downloaded your Python package. Beyond that, you assumed that most of those downloading your Python package did not have the competence as developers, programmers, coders, or simple hobbyists to parse the code before simply implementing it in their projects.~~

~~Or perhaps you felt that only someone like yourself would be uniquely qualified for that job.~~ 

~~In either case, you have a right to pull the code. That's not my qualm with you. My problem is with the **way you did it**. It was extremely insulting and, quite frankly, felt disrespectful to a certain extent.~~ 

~~You may have felt like you were doing some sort of good to the world by removing all of the code, but in my opinion it was an extremely pretentious decision. There are countless times that I've pulled some Python code to save myself time in recreating the basic structure of the package since that person's implementation had the basic gist of what I wanted.~~ 

~~Other times, I may just access code for the sole sake of extracting a *piece* of what they implemented.~~ 

~~For instance, maybe I know the cryptographic aspect of it, but I'm a little hazy on how one would connect to a certain backend or make API calls that authenticate the response, then pipe it into a variable or function (just random examples).~~ 

~~But I applaud you. Being the Samaritan you really are, you just decided that you were going to put your foot down and make sure all of us irresponsible children didn't get in over our heads with your excessively complex, unwieldy Python code that was guaranteed to be thrown into any project / implementation without a second thought or consideration.~~

**Comment by antoniusf** [comment]
Hi.

I assume that you're here because my change broke something on your end. I'm sorry for that.

I'm aware that I didn't handle the situation as well as I could have, and I'd appreciate concrete suggestions on what I could have done better. I don't have a lot of experience with this sort of thing; I'm happy to learn in case something like this comes up again.

You say that I pulled the code because I doubted the competence of the people who downloaded the package, because I thought they'd be unable to parse the code. I disagree: I pulled the code because this is a crypto library, and being able to understand the code does not mean that you know it is secure. Would you have been able to spot the nonce-reuse vulnerability while reading over the code? (I don't think I'd have – I found it by accident, and there might be other vulnerabilities that I simply don't know about.)

I wrote a library whose sole purpose it is to encrypt data so it can't be read by others. If people are using it, I think it's fair to assume they're using it to encrypt data they don't want others to see. I don't feel confident in providing that promise. I may be legally off the hook since the license has a very big notice that it comes with absolutely no warranty, but I'm still not comfortable with the thought that other people are relying on this.

In light of this, I can see why you might consider my behavior selfish, but I don't understand why you'd consider it arrogant. Please feel free to elaborate.

You also mentioned that I don't know what people might have been using the code for, and maybe this is right – from my point of view, there isn't much else you could do with it except for encrypting and decrypting things. So if you have some concrete examples, I'd be happy to hear them.

In any case: I'm sorry if this broke something for you. Feel free to let me know if I can do anything to help remedy the situation.

Thanks,
Antonius

**Comment by Librechain** [comment]
Hey, in light of your response - I agree with you. 

Rather than me forcing you to have to fix something that you may not have meant to release publicly, would you be interested in collaborating together to make this tool up to par with what you feel that users can safely implement on their end? 

I'm closing this issue btw, and modifying my stance on in light of your recent comment. I think "arrogant" was the wrong word to use. 

Perhaps a better way to express it is to say that I think you may not have considered the possibility that leaving the code out there 'in the wild' may have reeled in other intelligent minds such as yourself looking to help shore up the parts where you feel it is incomplete / inadequate for production use. 

Overall, the idea here is a good one. It would be a shame to not see it blossom to fruition (imo). 

But I'm sure that you are busy with your own personal projects, endeavors, and whatever else life has prepared for you, so I write this comment with optimism that you're still interested in tinkering with this project. 

Let me know how you feel about it.

**Comment by Librechain** [comment]
**Constructive Information** 

In case you are considering future collaboration on this project, I figured I would provide some information on what I was looking for when I stumbled across your tool. 

1. There is some sensitive information that I need to pipe from one socket / port to another **locally**. 
2. I know that there are other solutions like 'stunnel', but I was specifically looking for boilerplate code / module that would provide an ephemeral, encrypted stream into / from a file descriptor of some sort. Preferably something lightweight that could be called on & iterated in one script (before being destroyed). 

The main idea here was to provide an encrypted script for the transport of sensitive data that could be read in plain text by an adversary with access to the memory / RAM. Originally, I sought a solution that would allow me to clear the RAM / cache within the course of the script, but that still would leave my workflow vulnerable until said data was wiped. 

As time went on, I figured the best solution was essentially something like what you created - an encrypted "stream" of sorts that would allow me to 'cat' information to a file or pipe / re-direct info into file descriptors / sockets / unnamed pipes etc., ephemerally in a manner where, if an attacker were to gain the ability to access residual file traces in memory, they wouldn't be able to learn any useful info (i.e., commands, file locations, pre-images, etc.) 

I'm of the mind that anything that's being transcribed directly on command line (or executed via a readable program), is subject to forensic examination & thus, a potential source of information leak. 

Let me know your opinion if you get a chance

**Comment by antoniusf** [comment]
Hi,

thanks for your response, and explaining your motivation.

You're right that I'm currently quite busy – I've recently started a new job (and it's great!), which means that I have a lot less spare time to work on stuff.

I agree that it would be nice to see this turn into something proper, but I don't currently see how to do that without help from a person who actually knows what they're doing. I think the code is neat as a proof of concept, but it also has problems, some of them inherent in the structure of the problem it tries to solve. Overall, I currently don't see a good way to get something nice out of this.

However, the use case you described is a lot simpler than what the library currently does! One of the reasons I wrote this library, and also one of the reasons it's currently a bit too complex to be safe, is that it supports not just streaming data (linearly), but skipping around to arbitrary locations. In fact, if you only need to encrypt/decrypt streamed data, [libsodium has you covered!](https://libsodium.gitbook.io/doc/secret-key_cryptography/secretstream) This is awesome because it means that you don't have to touch cryptographically relevant code at all. Feel free to take a look!

As for your application, I'm not quite sure how much I can help. If you're piping data from one process to another process on the same machine, then all the data each of the processes use is also present in RAM, right? Including, for example, the encryption key for any encrypted data you might send across the socket? So defending against an adversary who has access to your RAM seems kind of difficult.

In either case, if you do want a library for encrypting/decrypting streams, why not take a look at libsodium's implementation?

Cheers,
Antonius
